app/models/cherrypick_task.rb

```diff
@@ -126,25 +132,85 @@ def add_empty_well(wells)
     private :add_empty_well
   end
 
-  def pick_new_plate(requests, template, robot, plate_purpose)
+  #
+  # Returns a list with the destination positions for the control wells distributed by
+  # using batch_id and num_plate as position generators.
+  def control_positions(batch_id, num_plate, num_free_wells, num_control_wells)
+    unique_number = batch_id
+
+    # Generation of the choice
+    positions = []
+    available_positions = (0..(num_free_wells - 1)).to_a
+
+    while positions.length < num_control_wells
+      current_size = available_positions.length
+      position = available_positions.slice!(unique_number % current_size)
+      position_for_plate = (position + num_plate) % num_free_wells
+      positions.push(position_for_plate)
+      unique_number /= current_size
+    end
+
+    positions
+  end
+
+  def pick_new_plate(requests, template, robot, plate_purpose, control_plate = nil)
     target_type = PickTarget.for(plate_purpose)
-    perform_pick(requests, robot) do
+    perform_pick(requests, robot, control_plate) do
       target_type.new(template, plate_purpose.try(:asset_shape))
     end
   end
 
-  def pick_onto_partial_plate(requests, template, robot, partial_plate)
+  def pick_onto_partial_plate(requests, template, robot, partial_plate, control_plate = nil)
     purpose = partial_plate.plate_purpose
     target_type = PickTarget.for(purpose)
 
-    perform_pick(requests, robot) do
+    perform_pick(requests, robot, control_plate) do
       target_type.new(template, purpose.try(:asset_shape), partial_plate).tap do
         partial_plate = nil # Ensure that subsequent calls have no partial plate
       end
     end
   end
 
-  def perform_pick(requests, robot)
+  # Creates control requests for the control assets provided and adds them to the batch
+  def create_control_requests!(batch, control_assets)
+    order = batch.requests.first.submission.orders.first
+    destination_control_wells = nil
+    order.create_requests_for_assets!(control_assets, nil, true) do |created_wells|
+      destination_control_wells = created_wells
+    end
+    requests_from_controls = destination_control_wells.map(&:requests_as_target).flatten
+    batch.requests << requests_from_controls
+
+    requests_from_controls
+  end
+
+  # Adds any consecutive list of control requests into the current_destination_plate
+  def add_any_consecutive_control_requests(control_positions, batch, control_assets, current_destination_plate)
+    while (asset_position = control_positions.find_index(current_destination_plate.content.length))
```

Comment:
Should this be a comparison  == not assignment = ?

Reply:
OK we think we understand it, it will stop when `control_positions.find_index(current_destination_plate.content.length)` returns nil, so asset_position is not truthy?

Reply:
Yes, so it assigns and checks condition in same line. I thought it was more tidy than assigning and then checking the condition.

Reply:
The whole method is very neat, but we did have some trouble understanding it :/

I wonder if something more verbose like this could be more readable? (Haven't tested it!)
```  
# Adds any consecutive list of control requests into the current_destination_plate
def add_any_consecutive_control_requests(control_positions, batch, control_assets, current_destination_plate)
  # find the index of the well we are filling right now
  current_well_index = current_destination_plate.content.length

  # check if this well should contain a control
  while control_positions.include?(current_well_index)
    control_asset = control_assets[control_positions.find_index(current_well_index)]
    add_control_request(batch, control_asset, current_destination_plate)
    # above adds to current_destination_plate, so current_well_index should be recalculated
    current_well_index = current_destination_plate.content.length
  end
end
```

Just a suggestion, see what you think. It is a lot longer

Reply:
Yes, is more descriptive and does the same thing. I'll update it.